### PR TITLE
Modifies the Mechfab to use the Ore Silo

### PR DESCRIPTION
**Pull request by TheVekter**
## About The Pull Request

I have wanted this for ages. The Mechfab will now pull from the Ore Silo as opposed to having to manually put materials inside. Ported from Yogstation, big ups to @alexkar598 for taking the time to figure this out on their end and make my life a lot easier and to @kriskog and @AnturK for testing and assistance in making it work with our materials datum.

## Why It's Good For The Game

I spent more time on this than I will ever save walking to R&D for materials and walking back and it's totally worth it

## Changelog
:cl:
add: Modifications have been made to the Mech Fabricator to allow it to withdraw and deposit to the ore silo (Thanks to alexkar598 from Yogstation for the code!)
/:cl:

### CODE REVIEW

**Comment by Fikou** [comment]
you can get mats from the circuit imprinter

**Comment by TheVekter** [comment]
> you can get mats from the circuit imprinter

I came back about a month ago and I have never, _ever_ had that work for me. I don't know why. Besides, it's the only machine in research that makes stuff and doesn't pull from the silo, so.

**Comment by tralezab** [comment]
and if that's true then there's no problem with this pr

**Comment by kriskog** [comment]
> 
> 
> > you can get mats from the circuit imprinter
> 
> I came back about a month ago and I have never, _ever_ had that work for me. I don't know why. 

Its a bug when you use it through the console, works fine if you use the printer directly. ¯\\\_(ツ)_/¯

**Comment by TheVekter** [comment]
That explains a lot. I'm still keeping this open - it's a QoL change that will have to happen eventually if the mech fabs are moved over to research.

**Comment by kriskog** [comment]
Oh yeah, for sure. Ore silo is a thing now, and most stuff should be connected to it. Autolathes too, if we are being honest.

**Comment by RaveRadbury** [comment]
I thought this and autolathes were separate for balance reasons, great QoL improvement otherwise.

**Comment by TheVekter** [comment]
I want to say that I think the autolathes are because they're much easier to access. I figure we can do a testmerge if we think it'll unbalance things. Otherwise, maybe we tweak material costs? I don't feel like "making it annoying" is a good way to balance something.

**Comment by Tharcoonvagh** [comment]
make it linkable to the silo too, not sure if you have already done this or not

**Comment by Tharcoonvagh** [comment]
also you can't fit that many sheets into the exofab, so this is a bit unbalanced i guess

**Comment by Detective-Google** [comment]
this will (help) stop roboticists hiding all the station's materials in their exofabs 👍

**Comment by wesoda25** [comment]
👎 bc I like watching roboticists suffer

**Comment by TheVekter** [comment]
> make it linkable to the silo too, not sure if you have already done this or not

Planned future feature right now. That's not something that's possible on Yog's code at the moment, so I'd have to hand script it, which I don't mind doing but will likely take a while. Though I'm tempted to say that you can't as a balance thing - it might actually be a problem if people manage to build and link another mechfab and can just shit out Durands as traitors.



> also you can't fit that many sheets into the exofab, so this is a bit unbalanced i guess

This is actually a good point, but like I said, I'd rather see how this pans out and adjust material costs than just balance this by making it mildly inconvenient.

**Comment by kriskog** [comment]
>
>
>  This is actually a good point

It's not. Storage size is a non issue. Having to juggle amount of each sheet is clunkiness, not """balance""".

Make it linkable to the ore silo. It requires going to the vault for the silo data, and you can rob any lathe way easier. Just steal the multitool code from protolathe/circuit printer.

**Comment by TheVekter** [comment]
> Make it linkable to the ore silo. It requires going to the vault for the silo data, and you can rob any lathe way easier. Just steal the multitool code from protolathe/circuit printer.

It's unfortunately not that easy. The ore silo and remote_materials.dm handle the linking and, for some reason, I can't get the code to actually work on the mech fab. I'll poke at it, but that's a low priority change for me right now. I'm more worried about figuring out why I can't pull from the mech fab but I'm tempted to just say "fuck it" and save that for later.

**Comment by TheVekter** [comment]
DNM, I have some bugs I need to sort out tonight.

File: code/game/mecha/mech_fabricator.dm [hunk 1]
```diff
@@ -144,8 +143,21 @@
 	desc = "It's building \a [initial(D.name)]."
 	var/list/res_coef = get_resources_w_coeff(D)
 
-	var/datum/component/material_container/materials = GetComponent(/datum/component/material_container)
+	var/datum/component/material_container/materials = rmat.mat_container
+	if (!materials)
+		say("No access to material storage, please contact the quartermaster.")
+		return 0
```
Comment (kriskog) [inline code/game/mecha/mech_fabricator.dm]: FALSE

File: code/game/mecha/mech_fabricator.dm [hunk 2]
```diff
@@ -144,8 +143,21 @@
...
 	var/list/res_coef = get_resources_w_coeff(D)
 
-	var/datum/component/material_container/materials = GetComponent(/datum/component/material_container)
+	var/datum/component/material_container/materials = rmat.mat_container
+	if (!materials)
+		say("No access to material storage, please contact the quartermaster.")
+		return 0
+	if (rmat.on_hold())
+		say("Mineral access is on hold, please contact the quartermaster.")
+		return 0
```
Comment (kriskog) [inline code/game/mecha/mech_fabricator.dm]: FALSE

File: code/game/mecha/mech_fabricator.dm [hunk 3]
```diff
@@ -397,10 +406,19 @@
...
 
-/obj/machinery/mecha_part_fabricator/on_deconstruction()
-	var/datum/component/material_container/materials = GetComponent(/datum/component/material_container)
-	materials.retrieve_all()
-	..()
+/obj/machinery/mecha_part_fabricator/proc/eject_sheets(eject_sheet, eject_amt)
+	var/datum/component/material_container/mat_container = rmat.mat_container
+	if (!mat_container)
+		say("No access to material storage, please contact the quartermaster.")
+		return 0
```
Comment (kriskog) [inline code/game/mecha/mech_fabricator.dm]: FALSE

File: code/game/mecha/mech_fabricator.dm [hunk 4]
```diff
@@ -397,10 +406,19 @@
...
+	if (rmat.on_hold())
+		say("Mineral access is on hold, please contact the quartermaster.")
+		return 0
```
Comment (kriskog) [inline code/game/mecha/mech_fabricator.dm]: FALSE

**Comment by FloranOtten** [comment]
Is the maintenance exofab on metastation connected at roundstart because of this? If it is, it'd be a bit wierd to be able to walk into a maint room and print off a mech (even if said mech is a shitty ripley).

File: code/game/mecha/mech_fabricator.dm (hunk 1 above)
Comment (TheVekter) [reply code/game/mecha/mech_fabricator.dm]: Fixed

File: code/game/mecha/mech_fabricator.dm (hunk 2 above)
Comment (TheVekter) [reply code/game/mecha/mech_fabricator.dm]: Fixed

**Comment by TheVekter** [comment]
> Is the maintenance exofab on metastation connected at roundstart because of this? If it is, it'd be a bit wierd to be able to walk into a maint room and print off a mech (even if said mech is a shitty ripley).

It does, but a Ripley takes a good chunk of iron and I'm not entirely certain one guy running around in a ripley without any of the good components is a huge problem. We can always just remove it from that map if we have to.

**Comment by Ty-the-Smonk** [comment]
Hasn’t this been denied before

**Comment by TheVekter** [comment]
> Hasn’t this been denied before

Nope, not that I could find. There was one that shoved all of the mechs into the departmental tech fabs, maybe that's the one you're thinking of?

**Comment by zxaber** [comment]
Not really a super huge fan of this, since I like to depend on the mats inside my ExoFabs not going anywhere. But then, I can just dysync them at round start so it probably won't negatively affect me too bad.

**Comment by TheVekter** [comment]
> Not really a super huge fan of this, since I like to depend on the mats inside my ExoFabs not going anywhere. But then, I can just dysync them at round start so it probably won't negatively affect me too bad.

That was my thought, too. You can just unscrew and rescrew it and it'll go back to the old behavior if you really want it.

> It's not really a great sign if a QoL change might possibly require removal of map features due to balance reasons. Why not just make a child object of the ExoFab that doesn't auto-sync at round start and stick that in maint?

I could, but mech code is really finnicky. The better option seems to be to have the mech fabs that start in maintenance be deconstructed, but place the parts nearby to put them back together. This way they have the old behavior without losing anything in exchange.

**Comment by kriskog** [comment]
Overriding the roundstart linking behaviour for the maint ones isn't really all that difficult, if you would rather do that instead of leaving them deconstructed. Basically add a subtype with the linking logic in Initialize, and put those in robotics.

**Comment by TheVekter** [comment]
This is good to merge.

**Comment by kriskog** [comment]
Delta also has a maint mechfab

**Comment by TheVekter** [comment]
> Delta also has a maint mechfab

Thanks, I'll take care of it.

**Comment by carlarctg** [comment]
make it take all ores except iron

**Comment by AnonymousNow** [comment]
Yes, please, for the love of god. Mechfabs have been left behind, have an old and shitty interface, and really need some love to be brought in line with other manufacturing machinery - this is one such step.

**Comment by TheVekter** [comment]
> Yes, please, for the love of god. Mechfabs have been left behind, have an old and shitty interface, and really need some love to be brought in line with other manufacturing machinery - this is one such step.

Someone (I can't remember who) is working on updating the UI to use tgui-next.

**Comment by AnonymousNow** [comment]
> 
> 
> > Yes, please, for the love of god. Mechfabs have been left behind, have an old and shitty interface, and really need some love to be brought in line with other manufacturing machinery - this is one such step.
> 
> Someone (I can't remember who) is working on updating the UI to use tgui-next.

Cracking. I swear they've been the same for nearly a decade, and I remember a time when they could technically print milk cream cartons (yes, it was a bug).

**Pull request merged**